### PR TITLE
[SLP][NFCI] Address issues seen in downstream Coverity scan.

### PR DESCRIPTION
**Pull request by tylanphear**
- Prevent null dereference: if the Mask given to
  `ShuffleInstructionBuilder::adjustExtracts()` is empty or all-poison,
  then `VecBase` will be `nullptr` and the call to
  `castToScalarTyElem(VecBase)` will dereference it. Add an assert
  to guard against this.

- Prevent use of uninitialized scalar: in the unlikely event that
  `CandidateVFs` is empty, then `AnyProfitableGraph` will be
  uninitialized in `if` condition following the loop. (This seems like a
  false-positive, but I submitted this change anyways as initializing
  bools costs nothing and is generally good practice)

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -11700,8 +11700,7 @@ class BoUpSLP::ShuffleInstructionBuilder final : public BaseShuffleAnalysis {
       R.eraseInstruction(EI);
     }
     if (NumParts == 1 || UniqueBases.size() == 1) {
-      VecBase = castToScalarTyElem(VecBase);
-      return VecBase;
+      return VecBase ? castToScalarTyElem(VecBase) : nullptr;
```
Comment (justinfargnoli) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:11703]: When not do this check in `castToScalarTyElem()`?

Comment (tylanphear) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:11703]: Hm, maybe it makes sense to add an `assert` there? I'm not sure `nullptr` is an expected input for this function, but I'm admittedly not too familiar with the SLP Vectorizer.

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:11703]: Add an assertion hete

Comment (tylanphear) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:11703]: Add an assertion in `castToScalarTyElem()`, at this callsite (L11703), or both? I think you mean at this callsite, but just wanted to be clear.

Comment (tylanphear) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: Added an assertion here at the callsite.

**Review by alexey-bataev — APPROVED** [review]
LG

**Pull request merged**